core/src/main/java/org/infinispan/container/offheap/ContiguousOffsetCalculator.java

```diff
@@ -36,6 +43,7 @@ public int calculateOffsetUsingHashCode(int hashCode) {
    }
 
    private static int spread(int h) {
-      return (h ^ (h >>> 16)) & HASH_BITS;
+      // Spread using murmur hash then ensure it is positive before finding block to use
+      return hash.hash(h) & HASH_BITS;
```

Comment:
Thinking about this some more, I am not sure if this how we want this to finally work. Unfortunately this means we are using the same spread as the distribution, which in DIST and REPL means that we will have a lot of entries that will probably map to close spread values. This might increase hit ratio internally.

Reply:
I wouldn't say it increases the hit ratio, because that sounds like a good thing, increasing the number of conflicts sounds a bit more accurate :) But I agree, if we use the same MH3 spread + division algorithm as `HashFunctionPartitioner` we're likely to end up with some empty buckets.

OTOH I think the modulo offset calculator should be just fine :)

Reply:
There's an interesting discussion of hash functions here: https://news.ycombinator.com/item?id=17328756

Reply:
Yeah I meant conflict (as in same address bucket).

I can revert this to use modulus offset and 9.4.1 can explore changing this. It should just require undoing the `entryStream` and references to modulus offset.

Reply:
I can then test specifically the different offset calculators (get and put should be same perf, but iteration _should_ be faster for contiguous).

Reply:
So the modulus implementation of `entryStream` currently prioritizes contiguous memory access to ensure CPU hit performance at the cost of lock acquisition and possible memory allocation.

Note https://github.com/infinispan/infinispan/pull/6249/commits/36bd8127e91c39568df2af61b20bae652f6e500c#diff-8234af2c969ad123972c741f990542bdR677 is invoked for every address pointer, thus having to acquire the `readLock` address pointer number of times (and having to create a Stream for every non 0 address pointer reference).

The contiguous version also reads in contiguous memory for address pointer, but since the locks are kept in contiguous order, we don't have to acquire a lock more than once and only one stream (in best case)
https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/container/offheap/OffHeapConcurrentMap.java#L681

In the case that address pointers are higher, which is the more common case it does batching of 256 address pointers at a time. So this reduces the lock acquisition and `Stream` allocation to be 1/256th of what it is with modulus.

https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/container/offheap/OffHeapConcurrentMap.java#L709

Obviously the 256 can be tweaked very easily in the file, hate magic numbers myself.

Reply:
Do you really need a lock to compute the address/bucket index?

If not, you could compute the bucket index outside the lock, and compute the lock index as `bucket index & (lock count - 1)` (assuming lock count is a power of 2).
Then for iteration you could have the same strategy of locking a range of buckets and creating a stream for all their entries.

Reply:
So the lock is required to ensure that the off heap memory is still available. If we don't acquire the lock we could crash the JVM if we had a concurrent stop when trying to access memory via the `memoryLookup` object.

Lock count is always a power of 2, and the same with address count. I had thought about locking a range of buckets with modulus, I tried it originally, but it gets ugly really fast as it isn't very conducive to the Stream API. I just sorta gave up. But I would be open to exploring it more as well, just maybe for 9.4.1 :)

Reply:
Definitely +1 to explore locking a contiguous range of buckets.